Reduce blastn TSV output to just the 6 fields parsed

peterjc opened the pull request:
This cuts the size of the intermediate TSV output by over half, so will hopefully have a small improvement on the runtime.

Note: Did not check-in the revised NJS files. The datastamp change is not of interest, but curiously the "bytes-total" values on macOS are much larger than on Linux. Perhaps this is linked to the file system?

Closes #127.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [x] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [ ] Fork the `pyani-plus` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [x] Set up an appropriate development environment (please see `CONTRIBUTING.md` for this repository)
- [x] Create a new branch with a short, descriptive name
- [x] Work on this branch
  - [x] style guidelines have been followed (please see `CONTRIBUTING.md` for this repository)
  - [x] new code is commented, especially in hard-to-understand areas
  - [x] corresponding changes to documentation have been made
  - [x] tests for the change have been added that demonstrate the fix or feature works
- [x] Test locally with `pytest-plus -v` **non-passing code will not be merged**
- [x] Rebase against `origin/master`
- [x] Check changes with linting/formatting tools before submission (please see `CONTRIBUTING.md` for this repository)
- [x] Commit branch
- [x] Submit pull request, describing the content and intent of the pull request
- [x] Request a code review
- [x] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani-plus/pulls) in the `pyan-plus` repository

peterjc commented on pyani_plus/methods/method_anib.py:
```diff
@@ -40,19 +40,18 @@
+BLAST_COLUMNS = ("qseqid pident length mismatch qlen gaps").split()
 
 # Precompute the column indexes once
+BLAST_COL_QUERY = BLAST_COLUMNS.index("qseqid")  # alternative to qaccver in std
```
It seems using either ``qseqid`` or ``qaccver`` will give the expected values like ``frag00001`` but hypothetically one could be fractionally faster than the other in BLAST's TSV output?

peterjc commented on pyani_plus/methods/method_anib.py:
```diff
@@ -121,7 +120,7 @@ def parse_blastn_file(blastn: Path) -> tuple[float | None, int | None, int | Non
                 )
                 raise ValueError(msg)
-            query = fields[0]
+            query = fields[BLAST_COL_QUERY]
```
This was implicitly assuming the query was the first column (column index zero).

widdowquinn commented:
Do we have a measure of performace gain from this?

It makes sense in principle not to store more information in the intermediate files than is necessary. As @kiepczi notes - what we consider "necessary" may not just be what is required for the results database, but also what is required to trace any issues in the analysis. We should keep that in mind.

What else do you think we might need in order to be able to debug failing/incorrect ANIb results quickly?

peterjc commented:
I think the pectobacterium test case I have run on Aphrodite with 200 genomes would be a good sized before/after comparison.

widdowquinn commented:
> I think the pectobacterium test case I have run on Aphrodite with 200 genomes would be a good sized before/after comparison (long enough to see any speedup).

Great - let's see how much time it buys us.

peterjc commented:
Rebased and updated to keep 7 columns (the 6 we need, plus the subject identifier).

Test case of 33 pecto:

```console
$ rm -rf x.db; touch x.db; time pyani-plus anib --database x.db pecto_md5_33/
```

Master branch 16m4s (and I had recorded a time of 15m56s prior to #216). This branch with seven columns:

```
$ rm -rf x.db; touch x.db; time pyani-plus anib --database x.db pecto_md5_33/
Indexing FASTAs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 33/33
ANIb run setup with 33 genomes in database
Database already has 0 of 33²=1089 ANIb comparisons, 1089 needed
host: n23-48-192-aphrodite
Comparing pairs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:15:45 1089/1089
Completed ANIb run-id 1 with 33 genomes in database x.db

real    15m48.571s
user    452m33.344s
sys     2m34.017s
```

So a fraction quicker but we'd need to repeat this a few times and/or use the larger test case to be confident.

Note that since #216, small examples like this (number of genomes < number of threads) no longer show any meaningful progress until it is nearly done (i.e. until the first column completes). I have some thoughts on that to follow up on later.

peterjc commented:
Larger test, with this branch and only 7 columns, using 48 cores and 200 genomes:

```console
$ rm -rf x.db; touch x.db; time pyani-plus anib --database x.db pecto_md5_200/
Indexing FASTAs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:04 200/200                       
ANIb run setup with 200 genomes in database
Database already has 0 of 200²=40000 ANIb comparisons, 40000 needed
host: n23-48-192-aphrodite
Comparing pairs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 8:02:44 40000/40000                   
Completed ANIb run-id 1 with 200 genomes in database x.db                                           

real    482m52.987s
user    20571m11.165s
sys     103m40.353s
```

Compared to a previous time of 488m22s, saved about 5.5 mins or about 1%.
